Add Jaszczak phantom

paolafer opened the pull request:
This PR adds the possibility of placing a Jaszczak phantom in the full body geometry. It is needed to do some of the NEMA studies.

paolafer committed:
Add Jaszczak phantom geometry

paolafer committed:
Modify full body geometry to use Jaszczak if needed

paolafer committed:
Rotate vector after translating

paolafer committed:
Extract function to build spheres

paolafer committed:
Improve code readability

paolafer committed:
Use auto when the variable type is clear

paolafer committed:
Add activities for the different parts

paolafer committed:
Add vertex generator

paolafer committed:
Add vertex in full body geometry

paolafer committed:
Cosmetics

carmenromo commented on source/geometries/FullRingInfinity.cc:
```diff
@@ -580,16 +581,20 @@ void FullRingInfinity::BuildSeparators()
+  jas_phantom_ = new JaszczakPhantom();
+  jas_phantom_->Construct();
+  G4LogicalVolume* phantom_logic = jas_phantom_->GetLogicalVolume();
+  new G4PVPlacement(0, G4ThreeVector(0, 0, 0), phantom_logic, "JASZCZAK", lab_logic_, false, 0, true);
```
Is it always placed in the center of the geometry?

paolafer replied:
I don't know, but if we need to translate it we can add the possibility later on. I think that the currents studies are being done with the phantom in the centre.

carmenromo commented on source/geometries/JaszczakPhantom.h:
```diff
@@ -0,0 +1,76 @@
+  G4GenericMessenger *msg_;
+  G4Navigator* geom_navigator_;
+
+  SpherePointSampler* spheric_gen_;
```
When compiling appears a warning message saying:

```
In file included from source/geometries/JaszczakPhantom.cc:9:
source/geometries/JaszczakPhantom.h:47:23: warning: private field 'spheric_gen_' is not used
```

paolafer replied:
Removed.

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+                                    rod4_d_(7.9*mm),
+                                    rod5_d_(9.5*mm),
+                                    rod6_d_(11.1*mm),
+                                    rod_height_(88*mm)
```
Which model have you use for the dimensions? I found for this company with two different phantoms but none of them uses `rod1_d_(3.2*mm)`.

https://capintec.com/product/jaszczak-spect-phantom/#tab-specifications

paolafer replied:
I've used the dimensions of the phantom that Andrew and Jacek are studying, which I think is the standard NEMA.

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+  auto cyl_name = "CYLINDER";
+  auto cylinder_solid =
+    new G4Tubs(cyl_name, 0, (cylinder_inner_diam_+2*cylinder_thickn_)/2,
+               (cylinder_height_+2*cylinder_thickn_)/2., 0, twopi);
```
In `cylinder_height_` is it not already included `cylinder_thickn_`? I understand that in the case of `cylinder_inner_diam_` it must be added since we are referring to the inner diameter, but in height I was not sure.

paolafer replied:
The arguments of `G4Tubs` are the semi-length and the radius of the cylinder. Since we are creating a cylinder made of aluminum and then placing water inside it, we need to make it slightly bigger than the water volume in every dimension, right? In this case `cylinder_height_` is the length of the water volume inside. If you think it makes it clearer, I can use `cylinder_internal_height_` as a name for the variable.

carmenromo replied:
I see, there's no need to change it!

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+
+  G4Material* aluminum = G4NistManager::Instance()->FindOrBuildMaterial("G4_Al");
+  auto cylinder_logic =
+    new G4LogicalVolume(cylinder_solid, aluminum, cyl_name);
```
Why have you chosen aluminum? I thought the material was a type of polymer/crystal, like PMMA.

paolafer replied:
Changed.

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+    auto diam = 2 * r;
+
+    // Sector displacement from centre, to accommodate gap between sectors
+    auto gap =  14.4 * mm;
```
Is this `gap` standard or you computed the distance to fit all rods?

paolafer replied:
I have translated Jacek's code to our simulations, I have assumed it was correct.

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+
+    // Sector displacement from centre, to accommodate gap between sectors
+    auto gap =  14.4 * mm;
+    auto dx  = gap * cos(pi/6);
```
Why is it `pi/6` instead of `pi/3` or `2*pi/6`?

paolafer replied:
Same as above.

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+    const auto Bx = 1.0, By = sqrt(3);
+    auto a = 0;
+    for (bool did_b=true ; did_b; a+=1) {
+      did_b = false;
```
It is not clear to me what `did_b ` is and how the two loops work.

paolafer replied:
Same as above.

carmenromo replied:
But you don't understand it either? I printed `did_b` and everytime it is `true` until line 162, but there is no condition for it, while `a` is increasing until it reaches line 167 and breaks. I still don't see the point of `did_b`.

paolafer replied:
This is how the `for (bool did_b=true; did_b; a+=1)` loop works: the first element inside the parenthesis is the starting value of the variable, the second element is the condition that keeps the loop going (meaning, when `did_b` is `false`, the loop stops), and the third element is executed in each iteration of the loop (meaning, `a` increases its value by 1). 

As far as I understand it, `did_b` becomes `false` at the beginning of each iteration of the outer loop. Then the inner loop starts, and at the end of each iteration `did_b` becomes `true` again. During the inner loop, the rods of each line are instantiated, and the loop breaks when it reaches the border of the cylinder or the other condition I added. When it breaks, we go back to the second iteration of the outer loop. If at least one rod has been instantiated, `did_b` is `true`, therefore the loop goes over the next iteration, because it fulfils the condition `did_b`. If not, the loop stops. This last case happens when in the inner loop no rods have been instantiated at all.

carmenromo replied:
Now I see, thank you!

carmenromo commented on source/geometries/JaszczakPhantom.cc:
```diff
@@ -0,0 +1,200 @@
+        G4ThreeVector pos = G4ThreeVector(x_pos, y_pos, z_pos);
+        G4ThreeVector newpos = pos.rotateZ(n*pi/3);
+        new G4PVPlacement(0, newpos, rod_logic, label, mother_logic, false, 0, true);
+
```
<img width="327" alt="Screenshot 2022-07-07 at 11 43 34" src="https://user-images.githubusercontent.com/32193826/177744344-1be407be-c0b5-42cb-99a1-63ae1deb0300.png">

Is the rod alone in the top right part on purpose?

paolafer replied:
I would say that the images that Andrew and Jacek sent us also had this. I think that it comes from the fact that the centres of the other corresponding rods are left outside the circle, therefore are not instantiated. If they fall inside, they are.

paolafer replied:
I have looked at this kind of phantoms in different companies' webpages and it seems that none of them have the configuration of the triangle with the smallest diameters. See, for instance, https://floridainstrumentacion.es/floridawp/wp-content/uploads/2017/12/Flanged_Jaszczak_Phantoms.pdf, ultra-deluxe model. It seems that the last row of rods is left out, right? I've checked the inner diameter of the cylinder and it's correct. I'll have a closer look and try to implement the geometry of this model.

paolafer replied:
Have a look now, and tell me what you think!

carmenromo replied:
> I have looked at this kind of phantoms in different companies' webpages and it seems that none of them have the configuration of the triangle with the smallest diameters.

Yes, that's what I saw yesterday and because of that I asked about the model.

>  It seems that the last row of rods is left out, right?

Yes...

carmenromo replied:
> Have a look now, and tell me what you think!

Perfect!

carmenromo reviewed (commented):
I would make also a test here to check for example that the number of rods, the materials and intensity are correct.

paolafer commented:
> I would make also a test here to check for example that the number of rods, the materials and intensity are correct.

How would you implement such a test?

paolafer committed:
Remove unused variable

paolafer committed:
Use plexiglass for the cylinder material

paolafer committed:
Displace slightly the rods from the centre to mimic Ultra-Deluxe

Ultra-Deluxe Jaszczak Phantom has been taken as a model. This change
makes the triangle with the smallest rods more similar to the original.

carmenromo commented:
> > I would make also a test here to check for example that the number of rods, the materials and intensity are correct.
> 
> How would you implement such a test?

mm I was thinking of asking for the correct number of rods and position, but I realized it is not like when we add sensors or wires, that we can check for ids.

carmenromo commented:
There's still something I don't understand. Gammas should be created in the phantom volume, more frequently in the rods and the spheres, aren't they? Because I simulated 1000 events and the initial position for the primary particles is always (0,0,0). Is it a matter of lack of statistics? It cannot be because even if they are always created in `WATER_BCKG` the position should change and not remain only in the center. Or does the center refer to the position of the phantom?

paolafer commented:
> There's still something I don't understand. Gammas should be created in the phantom volume, more frequently in the rods and the spheres, aren't they? Because I simulated 1000 events and the initial position for the primary particles is always (0,0,0). Is it a matter of lack of statistics? It cannot be because even if they are always created in `WATER_BCKG` the position should change and not remain only in the center. Or does the center refer to the position of the phantom?

Which macro are you using?

carmenromo commented:
`PET_full_body_nest`, the example for the Full body but setting the phantom to true.

paolafer committed:
Add example macros for this geometry

paolafer committed:
Remove overlap checks for rods ans spheres

paolafer commented:
> `PET_full_body_nest`, the example for the Full body but setting the phantom to true.

You have to use also the region corresponding to the phantom generation. I have added a macro as an example.

carmenromo commented on macros/PET_full_body_nest_jas.config.mac:
```diff
@@ -0,0 +1,40 @@
+
```
I would call the macros `PET_full_body_nest_phantom`, I think it's more understandable. And the output file `full_body_nest_phantom`, I lost my nest simulated file since they share name.

paolafer replied:
Done. However, take into account that these macros are just examples. If the user wants to create a file and keep it, they should take care of using an adequate name.

carmenromo commented:
A possible test would be simulating the phantom geometry and asking for the initial volumes of the primary particles, which can only be: `WATER_BCKG`, `ROD1`, `ROD5`, `ROD3`, `ROD0`, `ROD4`, `SPHERE4`.

paolafer committed:
Rename example macros

paolafer committed:
Add test to assert correct generation of vertices in phantom

carmenromo commented on tests/pytest/produce_output_test.py:
```diff
@@ -286,3 +286,70 @@ def test_create_nexus_output_file_nest(config_tmpdir, output_tmpdir, PETALODIR,
+
+     my_env    = os.environ
+     petalo_exe = PETALODIR + '/bin/petalo'
+     command   = [petalo_exe, '-b', '-n', '10000', init_path]
```
Last thing, I promise! :) Is it not enough with 100 evts?

paolafer replied:
No problem, that's the purpose of a review :-). The larger the statistics, the better to check this, so I decided to go for > 1000, and it's quite fast anyway, since this job doesn't generate optical photons.

carmenromo replied:
Ah, I didn't realized it does not generate optical photons. Then I approve!

carmenromo commented:
Possibility of adding the Jaszczak phantom in the full body geometry with its corresponding test has been included. Approved!
